neurom/view/_dendrogram.py

```diff
@@ -27,280 +27,165 @@
 
-from copy import deepcopy
-import sys
 import numpy as np
 
-from neurom.core import Tree, Neurite
+from neurom import NeuriteType
+from neurom.core import Neurite
 from neurom.core.dataformat import COLS
 
 
-def _n_terminations(tree):
-    '''Get the number of terminations in a tree'''
-    return sum(1 for _ in tree.ileaf())
+class Dendrogram:
+    '''Dendrogram'''
 
+    def __init__(self, neurom_section, dendrogram_root=None):
+        '''Dendrogram for NeuroM section tree.
 
-def _max_recursion_depth(obj):
-    ''' Estimate recursion depth, which is defined as the number of nodes in a tree
-    '''
-    neurites = obj.neurites if hasattr(obj, 'neurites') else [obj]
-
-    return max(sum(1 for _ in neu.iter_sections()) for neu in neurites)
-
-
-def _total_rectangles(tree):
-    '''
-    Calculate the total number of segments that are required
-    for the dendrogram. There is a vertical line for each segment
-    and two horizontal line at each branching point
-    '''
-    return sum(len(sec.children) + sec.points.shape[0] - 1
-               for sec in tree.iter_sections())
-
-
-def _n_rectangles(obj):
-    '''
-    Calculate the total number of rectangles with respect to
-    the type of the object
-    '''
-    return sum(_total_rectangles(neu) for neu in obj.neurites) \
-        if hasattr(obj, 'neurites') else _total_rectangles(obj)
-
-
-def _square_segment(radius, origin):
-    '''Vertices for a square
-    '''
-    return np.array(((origin[0] - radius, origin[1] - radius),
-                     (origin[0] - radius, origin[1] + radius),
-                     (origin[0] + radius, origin[1] + radius),
-                     (origin[0] + radius, origin[1] - radius)))
-
-
-def _vertical_segment(old_offs, new_offs, spacing, radii):
-    '''Vertices for a vertical rectangle
-    '''
-    return np.array(((new_offs[0] - radii[0], old_offs[1] + spacing[1]),
-                     (new_offs[0] - radii[1], new_offs[1]),
-                     (new_offs[0] + radii[1], new_offs[1]),
-                     (new_offs[0] + radii[0], old_offs[1] + spacing[1])))
-
-
-def _horizontal_segment(old_offs, new_offs, spacing, diameter):
-    '''Vertices of a horizontal rectangle
+        Args:
+            neurom_section: NeuroM section tree.
+            dendrogram_root: root of dendrogram. This is a service arg, please don't set it on
+            your own. It is used to track cycles in ``neurom_section``.
+        '''
+        if dendrogram_root is None:
+            dendrogram_root = self
+            dendrogram_root.processed_section_ids = []
+        if neurom_section.id in dendrogram_root.processed_section_ids:
+            raise ValueError('Cycled morphology {}'.format(neurom_section))
+        dendrogram_root.processed_section_ids.append(neurom_section.id)
+
+        segments = neurom_section.points
+        segment_lengths = np.linalg.norm(
+            np.subtract(segments[:-1, COLS.XYZ], segments[1:, COLS.XYZ]), axis=1)
```

Comment:
Do you mean to save it as a property? `this.segments_lengths = segment_lengths`? Currently I don't see any usage for it later.

Reply:
I mean use the function that does this already `neurom/fst/_neuritefunc.py`, but called using `get()` instead of directly

Reply:
Ah, ok. I will do the changes after all suggestions have been made.

Reply:
Unfortunately I couldn't reuse `segments_lengths` function because it is for neurites. I have sections. Please look [neurom/fst/_neuritefunc.py#L204](https://github.com/BlueBrain/NeuroM/blob/master/neurom/fst/_neuritefunc.py#L204).